controller/sync_tasks.go

```diff
@@ -0,0 +1,127 @@
+package controller
+
+import (
+	"strings"
+
+	. "github.com/argoproj/argo-cd/pkg/apis/application/v1alpha1"
+)
+
+// kindOrder represents the correct order of Kubernetes resources within a manifest
+var syncPhaseOrder = map[SyncPhase]int{
+	SyncPhasePreSync:  -1,
+	SyncPhaseSync:     0,
+	SyncPhasePostSync: 1,
+}
+
+// kindOrder represents the correct order of Kubernetes resources within a manifest
+var kindOrder = map[string]int{
+	"Namespace":                -26,
+	"ResourceQuota":            -24,
+	"LimitRange":               -24,
+	"PodSecurityPolicy":        -23,
+	"Secret":                   -22,
+	"ConfigMap":                -21,
+	"StorageClass":             -20,
+	"PersistentVolume":         -19,
+	"PersistentVolumeClaim":    -18,
+	"ServiceAccount":           -17,
+	"CustomResourceDefinition": -16,
+	"ClusterRole":              -15,
+	"ClusterRoleBinding":       -14,
+	"Role":                     -13,
+	"RoleBinding":              -12,
+	"Service":                  -11,
+	"DaemonSet":                -10,
+	"Pod":                      -9,
+	"ReplicationController":    -8,
+	"ReplicaSet":               -7,
+	"Deployment":               -6,
+	"StatefulSet":              -5,
+	"Job":                      -4,
+	"CronJob":                  -3,
+	"Ingress":                  -2,
+	"APIService":               -1,
+}
+
+type syncTasks []*syncTask
+
+func (s syncTasks) Len() int {
+	return len(s)
+}
+
+func (s syncTasks) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+// order is
+// 1. phase
+// 2. wave
+// 3. kind
+// 4. name
+func (s syncTasks) Less(i, j int) bool {
+
+	tA := s[i]
+	tB := s[j]
+
+	d := syncPhaseOrder[tA.phase] - syncPhaseOrder[tB.phase]
+	if d != 0 {
+		return d < 0
+	}
+
+	d = tA.wave() - tB.wave()
+	if d != 0 {
+		return d < 0
+	}
+
+	a := tA.obj()
+	b := tB.obj()
+
+	// we take advantage of the fact that if the kind is not in the kindOrder map,
+	// then it will return the default int value of zero, which is the highest value
+	d = kindOrder[a.GetKind()] - kindOrder[b.GetKind()]
```

Comment:
I'm not confident this logic handles the bug that we discovered (and fixed in upstream helm) where both resource kinds are unknown. @alexmt should look.

Reply:
Looks like it introduces the same bug again. CRDs should be synced last. @alexec please check unit test from this PR: https://github.com/argoproj/argo-cd/pull/1042/files